src/lib/tuning/tuning_choice.hpp

```diff
@@ -0,0 +1,128 @@
+#pragma once
+
+#include <memory>
+#include <set>
+
+#include "tuning/tuning_operation.hpp"
+
+namespace opossum {
+
+/**
+ * A TuningChoice represents a possible system modification with a certain
+ * performance impact coming at a cost.
+ *
+ * TuningChoices are generated by AbstractEvaluators in a Tuner instance
+ * and either accepted or rejected by an AbstractSelector while generating
+ * a sequence of TuningOperations to improve the overall system performance.
+ */
+class TuningChoice {
+ public:
+  virtual ~TuningChoice() {}
+  /**
+   * An estimate regarding the performance impact of this modification.
+   * values < 0.0f: expected degradation of system performance
+   * values > 0.0f: expected improvement of system performance
+   *
+   * No scaling constraints are specified so that this value is only comparable
+   * among TuningChoices generated by compatible AbstractEvaluator implementations.
+   */
+  virtual float desirability() const = 0;
+
+  /**
+   * How certain the producing evaluator was when generating this choice.
+   * Very basic evaluators will probably output choices with low confidence
+   * in general.
+   * Similarly, if an evaluator is very specialized for a certain edge case,
+   * it might produce few choices with high confidence and several other
+   * choices with a low confidence value, as other evaluators might be able
+   * to give better results in these cases.
+   */
+  virtual float confidence() const = 0;
+
+  /**
+   * Convenience accessors for desirability() that in their default implementation
+   * return the performance benetfts expected from accepting or rejecting this choice.
+   */
+  virtual float accept_desirability() const;
+  virtual float reject_desirability() const;
+
+  /**
+   * An estimate of the absolute costs of this modification.
+   *
+   * As this value is intended to be counted against a common budget,
+   * care must be taken that all AbstractEvaluators use the same cost measure.
+   */
+  virtual float cost() const = 0;
+
+  /**
+   * Convenience accessors for cost() that in their default implementation
+   * return the cost this choice currently imposes on the system and the respective
+   * cost delta resulting from an accept and reject operation.
+   */
+  virtual float current_cost() const;
+  virtual float accept_cost() const;
+  virtual float reject_cost() const;
+
+  /**
+   * True, if this modification is already present in the current system state.
+   */
+  virtual bool is_currently_chosen() const = 0;
+
+  /**
+   * A list of other TuningChoices that should/can not be chosen if this
+   * TuningChoice is accepted.
+   */
+  const std::set<std::shared_ptr<TuningChoice>>& invalidates() const;
```

Comment:
This is a bit different from `confidence` because you evaluator is already checking for this. Still, I would add a comment saying that no `invalidates`-relationships are currently produced

Reply:
This comment would need to be added to the IndexEvaluator, because the TuningChoice can not know how it is used by specific Evaluators.